docs/architecture/adr-018-dynamic-capability-store.md

```diff
@@ -0,0 +1,197 @@
+# ADR 18: Dynamic Capability Store
+
+## Changelog
+
+- 12 December 2019: Initial version
+
+## Context
+
+Full implementation of the [IBC specification](https://github.com/cosmos/ics) requires the ability to create and authenticate object-capability keys at runtime (i.e., during transaction execution),
+as described in [ICS 5](https://github.com/cosmos/ics/tree/master/spec/ics-005-port-allocation#technical-specification). In the IBC specification, capability keys are created for each newly initialised
+port & channel, and are used to authenticate future usage of the port or channel. Since channels and potentially ports can be initialised during transaction execution, the state machine must be able to create
+object-capability keys at this time.
+
+At present, the Cosmos SDK does not have the ability to do this. Object-capability keys are currently pointers (memory addresses) of `sdk.StoreKey` structs created at application initialisation in `app.go` ([example](https://github.com/cosmos/gaia/blob/master/app/app.go#L132))
+and passed to Keepers as fixed arguments ([example](https://github.com/cosmos/gaia/blob/master/app/app.go#L160)). Keepers cannot create or store capability keys during transaction execution — although they could call `sdk.NewKVStoreKey` and take the memory address
+of the returned struct, storing this in the Merklised store would result in a consensus fault, since the memory address will be different on each machine (this is intentional — were this not the case, the keys would be predictable and couldn't serve as object capabilities).
+
+Keepers need a way to keep a private map of store keys which can be altered during transacton execution, along with a suitable mechanism for regenerating the unique memory addresses (capability keys) in this map whenever the application is started or restarted.
+This ADR proposes such an interface & mechanism.
+
+## Decision
+
+The SDK will include a new `CapabilityKeeper` abstraction, which is responsible for provisioning, tracking, and authenticating capabilities at runtime. During application initialisation in `app.go`, the `sdk.CapabilityKeeper` will
+be hooked up to modules through unique function references so that it can identify the calling module when later invoked. When the initial state is loaded from disk, the `sdk.CapabilityKeeper` instance will create new capability keys
```

Comment:
```
When the initial state is loaded from disk, the `sdk.CapabilityKeeper` instance will create new capability keys
```

Can you elaborate a bit or make it somewhat more clear exactly how replay from genesis will work since cap-objects are address spaces?

Reply:
This is the `Initialise` function, I tried to clarify & referenced the function name.